File: oj/leetcode/2019/word-ladder/main.py
from typing import List


class Solution:
    def ladderLength(
        self,
        beginWord: str,
        endWord: str,
        wordList: List[str],
    ) -> int:
        from collections import deque

        if len(wordList) < 1:
            return 0

        wordList = set(wordList)

        def can_convert(a, b):
            diff_count = 0
            for ac, bc in zip(a, b):
                if ac != bc:
                    diff_count += 1
                    if diff_count > 1:
                        return False
            return diff_count == 1

        visited_words = set()
        pending = deque([
            (word, 2)
            for word in wordList
            if can_convert(word, beginWord)
        ])
        if beginWord in wordList:
            # special case
            pending.appendleft((beginWord, 1))
        while len(pending) > 0:
            word, distance = pending.popleft()
            visited_words.add(word)
            if word == endWord:
                return distance
            for new_word in wordList:
                if new_word in visited_words:
                    continue
                if not can_convert(new_word, word):
                    continue
                if new_word == endWord:
                    return distance + 1
                pending.append((new_word, distance + 1))

        return 0

File: oj/leetcode/2019/word-ladder/main.go
package main

import "fmt"

func ladderLength(beginWord string, endWord string, wordList []string) int {
	if len(wordList) < 1 {
		return 0
	}

	dict := map[string]struct{}{}
	for _, word := range wordList {
		dict[word] = struct{}{}
	}
	if _, ok := dict[endWord]; !ok {
		// special case
		return 0
	}

	pathLength := 1
	// each pass process one *whole* level
	shortPending := map[string]struct{}{}
	shortPending[beginWord] = struct{}{}
	longPending := map[string]struct{}{}
	longPending[endWord] = struct{}{}
	visitedWords := map[string]struct{}{}
	for len(shortPending) > 0 && len(longPending) > 0 {
		// O(n^d) >> O(n^(d/2) + n^(d/2))
		// so it's better to search both end (level search)
		if len(shortPending) > len(longPending) {
			// each pass, start from the shorter end
			longPending, shortPending = shortPending, longPending
		}

		// generate new level
		newPendingWords := map[string]struct{}{}
		// process this level
		for word, _ := range shortPending {
			visitedWords[word] = struct{}{}
			for i := 0; i < len(word); i++ {
				for c := 'a'; c <= 'z'; c++ {
					newWord := word[:i] + string(c) + word[i+1:]
					if _, valid := dict[newWord]; !valid {
						continue
					}
					if newWord == word {
						continue
					}
					if _, overlapped := longPending[newWord]; overlapped {
						// two ends meet, return
						return pathLength + 1
					}
					if _, visited := visitedWords[newWord]; visited {
						continue
					}
					newPendingWords[newWord] = struct{}{}
				}
			}
		}
		shortPending = newPendingWords
		pathLength += 1
	}

	return 0
}

func main() {
	if d := ladderLength("a", "c", []string{"a", "b", "c"}); d != 2 {
		panic(fmt.Sprintf("expected 2, got: %d", d))
	}
	fmt.Println("====")
	if d := ladderLength("a", "c", []string{"b", "c"}); d != 2 {
		panic(fmt.Sprintf("expected 3, got: %d", d))
	}
	if d := ladderLength("hot", "dog", []string{"hot", "dog"}); d != 0 {
		panic(fmt.Sprintf("expected 1, got: %d", d))
	}
}
